pallets/api/src/messaging/transports/ismp.rs

```diff
@@ -164,90 +156,120 @@ impl<T: Config> IsmpModule for Handler<T> {
-				let id = match request {
-					Get(get) => H256::from(keccak_256(&get.encode())),
-					Post(post) => H256::from(keccak_256(&post.encode())),
-				};
-				let key =
-					IsmpRequests::<T>::get(id).ok_or(Error::Custom("request not found".into()))?;
-				Messages::<T>::try_mutate(key.0, key.1, |message| {
-					let Some(super::super::Message::Ismp { commitment, deposit, .. }) = message
-					else {
-						return Err(Error::Custom("message not found".into()))
-					};
-					*message = Some(super::super::Message::IsmpTimedOut {
-						deposit: *deposit,
-						commitment: *commitment,
-					});
-					Ok(())
-				})?;
-				Ok(())
+				let commitment = hash_request::<T::Keccak256>(&request);
+				timeout_commitment::<T>(&commitment)
 			},
-			Timeout::Response(_response) => {
-				todo!()
+			Timeout::Response(PostResponse { post, .. }) => {
+				let commitment = hash_request::<T::Keccak256>(&Request::Post(post));
+				timeout_commitment::<T>(&commitment)
 			},
 		}
 	}
 }
 
-// TODO: replace with benchmarked weight functions
 impl<T: Config> IsmpModuleWeight for Pallet<T> {
+	// Static as not in use.
 	fn on_accept(&self, _request: &PostRequest) -> Weight {
-		todo!()
+		DbWeightOf::<T>::get().reads_writes(1, 1)
 	}
 
-	fn on_timeout(&self, _timeout: &Timeout) -> Weight {
-		DbWeightOf::<T>::get().reads_writes(2, 1)
+	fn on_timeout(&self, timeout: &Timeout) -> Weight {
+		let x = match timeout {
+			Timeout::Request(Request::Post(_)) => 0u32,
+			Timeout::Request(Request::Get(_)) => 1u32,
+			Timeout::Response(_) => 2u32,
+		};
+		T::WeightInfo::ismp_on_timeout(x)
 	}
 
-	fn on_response(&self, _response: &Response) -> Weight {
-		DbWeightOf::<T>::get().reads_writes(2, 2)
-	}
-}
-
-fn calculate_deposit<T: Config>(mut deposit: BalanceOf<T>) -> BalanceOf<T> {
-	// Add amount for `IsmpRequests` lookup.
-	let key_len: BalanceOf<T> =
-		(T::AccountId::max_encoded_len() + MessageId::max_encoded_len()).saturated_into();
-	deposit.saturating_accrue(
-		T::ByteFee::get() * (H256::max_encoded_len().saturated_into::<BalanceOf<T>>() + key_len),
-	);
+	// todo: test
+	fn on_response(&self, response: &Response) -> Weight {
+		let x = match response {
+			Response::Get(_) => 1,
+			Response::Post(_) => 0,
+		};
 
-	deposit
+		T::WeightInfo::ismp_on_response(x).saturating_add(T::CallbackExecutor::execution_weight())
+		// Also add actual weight consumed by contract env.
+	}
 }
 
-fn process_response<T: Config>(
+pub(crate) fn process_response<T: Config>(
 	commitment: &H256,
-	encode: &impl Encode,
-	store: impl Fn() -> Vec<u8>,
+	response_data: &impl Encode,
 	event: impl Fn(AccountIdOf<T>, MessageId) -> Event<T>,
 ) -> Result<(), anyhow::Error> {
-	let (origin, id) =
-		IsmpRequests::<T>::get(commitment).ok_or(Error::Custom("request not found".into()))?;
+	ensure!(
+		response_data.encoded_size() <= T::MaxResponseLen::get() as usize,
+		Error::Custom("Response length exceeds maximum allowed length.".into())
```

Comment:
How will the contract know about this? Should it be emitted as an event instead for offchain workflow rather than error? Perhaps messages should transition to errored as well as timed out?